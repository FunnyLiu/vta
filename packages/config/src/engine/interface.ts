/* eslint-disable @typescript-eslint/no-explicit-any */
export declare interface Config {
  [key: string]: any;
}
export declare interface ConfigByHelper<T> {
  type: symbol;
  payload: T;
}

export declare type Helper<T> = (store: Store, key: string, payload: T) => any;

export declare interface Helpers {
  has(type: symbol): boolean;
  regist<T>(type: symbol, Helper: Helper<T>): void;
  resolve(store: Store, key: string, type: symbol, payload: any): any;
}

export declare type EventListener = (...args: any[]) => void;

export declare interface Events {
  on(type: string, listener: EventListener): void;
  emit(type: string, ...args: any[]): void;
}

export declare interface Store {
  /**
   * regist config dir. baseMode=false is always last one
   * @param dir config dir
   * @param baseMode user's config please set false. default true
   */
  registDir(dir: string, baseMode?: boolean): void;
  getItem(key: string): Config;
  load(key: string): Config;
  reset(): void;
  events: Events;
  helpers: Helpers;
}
