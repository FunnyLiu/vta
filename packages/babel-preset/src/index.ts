import { resolveConfig } from "vta";

export default function vtaBabelPreset() {
  const { presets = [], plugins = [] } = resolveConfig("babel") || {};
  return { presets, plugins };
}
