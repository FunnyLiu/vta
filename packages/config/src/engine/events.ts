import { EventListener, Events } from "./interface";

export default class ConfigEvents implements Events {
  constructor() {
    this.listeners = {};
  }

  private listeners: { [type: string]: Array<EventListener> };

  public on(type: string, listener: EventListener) {
    let listeners = this.listeners[type];
    if (!listeners) {
      listeners = [];
      this.listeners[type] = listeners;
    }
    listeners.push(listener);
  }

  public emit<T>(type: string, ...args) {
    const listeners = this.listeners[type];
    if (listeners) {
      listeners.forEach((listener) => {
        listener(...args);
      });
    }
  }
}
