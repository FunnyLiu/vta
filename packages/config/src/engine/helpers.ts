import { Helper, Helpers, Store } from "./interface";

/* eslint-disable @typescript-eslint/no-explicit-any */

export default class ConfigHelpers implements Helpers {
  #helpers: Map<symbol, Helper<any>> = new Map();

  public has(type: symbol) {
    return this.#helpers.has(type);
  }

  public regist<T>(type: symbol, creator: Helper<T>) {
    this.#helpers.set(type, creator);
  }

  public resolve(store: Store, key: string, type: symbol, payload) {
    const helper = this.#helpers.get(type);
    return helper(store, key, payload);
  }
}
