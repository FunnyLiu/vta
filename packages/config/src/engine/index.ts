import { Helper, Store, Config } from "./interface";
import ConfigStore from "./store";
import ConfigHelpers from "./helpers";

const helpers = new ConfigHelpers();
const stores = new Map<string, Store>();
function getStore(category = "default") {
  if (!stores.has(category)) {
    stores.set(category, new ConfigStore(helpers));
  }
  return stores.get(category);
}
/* eslint-disable import/prefer-default-export */

/**
 * regist config helper
 * @param type helper type
 * @param helper receive store:Store,key:string,payload:T return anything
 */
export function registHelper<T>(type: symbol, helper: Helper<T>) {
  helpers.regist(type, helper);
}

/**
 * regist config dir. baseMode=false is always last one
 * @param dir config dir
 * @param baseMode user's config please set false. default true
 * @param category config category
 */
export function registDir(dir: string, baseMode = true, category?: string) {
  const store = getStore(category);
  store.registDir(dir, baseMode);
}

/* eslint-disable @typescript-eslint/no-explicit-any */

/**
 * get the sepecific key's config
 * @param key config key.`${key}.config.js`
 * @param category config category
 */
export function resolveConfig<T = Config>(key: string, category?: string): T {
  const store = getStore(category);
  return store.load(key) as any;
}

/**
 * clear store
 * @param category config category
 */
export function clearStore(category?: string) {
  const store = getStore(category);
  store.clear();
}

export const hooks = {
  onConfigBaseStart(key: string, cb: (store: Store) => void, category?: string) {
    const store = getStore(category);
    store.events.on(`config-${key}-base-start`, () => {
      cb(store);
    });
  },
  onConfigBaseDone(key: string, cb: (config: Config, store: Store) => void, category?: string) {
    const store = getStore(category);
    store.events.on(`config-${key}-base-done`, config => {
      cb(config, store);
    });
  },
  onConfigUserStart(key: string, cb: (store: Store) => void, category?: string) {
    const store = getStore(category);
    store.events.on(`config-${key}-user-start`, () => {
      cb(store);
    });
  },
  onConfigUserGetted(
    key: string,
    cb: (mergedConfig: Config, config: Config, store: Store) => void,
    category?: string,
  ) {
    const store = getStore(category);
    store.events.on(`config-${key}-user-getted`, (mergedConfig, config) => {
      cb(mergedConfig, config, store);
    });
  },
  onConfigUserDone(key: string, cb: (config: Config, store: Store) => void, category?: string) {
    const store = getStore(category);
    store.events.on(`config-${key}-user-done`, config => {
      cb(config, store);
    });
  },
};

/**
 * set the config's ext. default is js. only use for jest
 * @param ext
 */
export function setStoreExt(ext = "js", category?: string) {
  const store: ConfigStore = getStore(category) as ConfigStore;
  store.ext = ext;
}
