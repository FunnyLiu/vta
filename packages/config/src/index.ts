export { clearStore, resolveConfig, registDir, registHelper, hooks } from "./engine/index";
export { default as useBase } from "./helpers/use-base";
export { default as useDeps } from "./helpers/use-deps";
