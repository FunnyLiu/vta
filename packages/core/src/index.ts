export { default as run, appRunSync as runSync } from "./core/index";
export { Plugin, App, PrepareHelpers } from "./core/interface";
export { default as resolveConfig } from "./core/resolveConfig";
