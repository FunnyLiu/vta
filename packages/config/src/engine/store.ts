import path from "path";
import { deepMerge } from "@vta/helpers";
import { Config, Store, Events, Helpers } from "./interface";
import ConfigEvents from "./events";
import loadModule from "./utils/load-module";

export default class ConfigStore implements Store {
  constructor(helpers: Helpers) {
    this.helpers = helpers;
  }

  public ext = "js";

  private config: { [key: string]: Config } = {};

  private loadingKeys: Set<string> = new Set();

  private dirs: Array<{ dir: string; baseMode: boolean }> = [];

  public events: Events = new ConfigEvents();

  public helpers: Helpers;

  public registDir(dir: string, baseMode = true) {
    if (this.dirs.filter(({ dir: tempDir }) => tempDir === dir).length === 0) {
      const lastDir = this.dirs[this.dirs.length - 1];
      if (!lastDir || lastDir.baseMode) {
        this.dirs.push({ dir, baseMode });
      } else {
        this.dirs.splice(this.dirs.length - 1, 0, { dir, baseMode });
      }
    }
  }

  public getItem(key: string) {
    return this.config[key] || {};
  }

  private setItem(key: string, config: Config) {
    this.config[key] = config || {};
  }

  private resolveValue(key: string, value) {
    if (Object.prototype.toString.call(value) === "[object Object]") {
      if (typeof value.type === "symbol" && this.helpers.has(value.type)) {
        return this.resolveValue(key, this.helpers.resolve(this, key, value.type, value.payload));
      }
      /* eslint-disable no-param-reassign */
      Object.keys(value).forEach(prop => {
        value[prop] = this.resolveValue(key, value[prop]);
      });
      Reflect.ownKeys(value).forEach(prop => {
        value[prop] = this.resolveValue(key, value[prop]);
      });
    }
    if (Array.isArray(value)) {
      return value.map(v => this.resolveValue(key, v));
    }
    return value;
  }

  public load(key: string): Config {
    if (this.loadingKeys.has(key)) return this.getItem(key);
    this.loadingKeys.add(key);

    let userModeDir = this.dirs[this.dirs.length - 1];
    if (userModeDir.baseMode) userModeDir = undefined;
    const baseModelDirs = userModeDir ? this.dirs.slice(0, this.dirs.length - 1) : this.dirs;

    const mergeConfig = (config: Config) => {
      if (config) {
        this.setItem(key, deepMerge(this.getItem(key), this.resolveValue(key, config)));
      }
    };

    this.events.emit(`config-${key}-base-start`, mergeConfig);
    baseModelDirs.forEach(({ dir }) => {
      const config = deepMerge(
        this.getItem(key),
        this.resolveValue(
          key,
          loadModule<Config>(path.resolve(dir, `${key}.config.${this.ext}`), {}),
        ),
      );
      this.setItem(key, config);
    });
    this.events.emit(`config-${key}-base-done`, this.getItem(key), mergeConfig);
    this.events.emit(`config-${key}-user-start`, mergeConfig);
    if (userModeDir) {
      const { dir } = userModeDir;
      const config = deepMerge(
        this.getItem(key),
        this.resolveValue(
          key,
          loadModule<Config>(path.resolve(dir, `${key}.config.${this.ext}`), {}),
        ),
      );
      this.setItem(key, config);
    }
    this.events.emit(`config-${key}-user-getted`, this.getItem(key), mergeConfig);
    this.events.emit(`config-${key}-user-done`, this.getItem(key), mergeConfig);
    return this.getItem(key);
  }
}
