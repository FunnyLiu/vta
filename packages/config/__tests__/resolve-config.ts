import path from "path";
import { resolveConfig, registDir, clearStore } from "../src";
import { setStoreExt, hooks } from "../src/engine";

setStoreExt("ts");

const dir1 = path.resolve(__dirname, "./data/project-1/config");
const dir2 = path.resolve(__dirname, "./data/project-2/config");
const dir3 = path.resolve(__dirname, "./data/project-3/config");

describe("config", () => {
  it("hooks", () => {
    const category = "default";
    registDir(dir1);
    registDir(dir2);
    registDir(dir3, false);

    const processRecords = [];
    ["app", "env", "webpack", "webpack-server"].forEach(key => {
      hooks.onConfigBaseStart(
        key,
        () => {
          processRecords.push(`${key}-base-start`);
        },
        key === "app" ? category : undefined,
      );
      hooks.onConfigBaseDone(
        key,
        () => {
          processRecords.push(`${key}-base-done`);
        },
        key === "app" ? category : undefined,
      );
      hooks.onConfigUserStart(
        key,
        () => {
          processRecords.push(`${key}-user-start`);
        },
        key === "app" ? category : undefined,
      );
      hooks.onConfigUserGetted(
        key,
        () => {
          processRecords.push(`${key}-user-getted`);
        },
        key === "app" ? category : undefined,
      );
      hooks.onConfigUserDone(
        key,
        () => {
          processRecords.push(`${key}-user-done`);
        },
        key === "app" ? category : undefined,
      );
    });
    resolveConfig("env"); // env deps app
    resolveConfig("webpack"); // webpack deps app env webpack-server
    expect(JSON.stringify(processRecords, null, 2)).toMatchSnapshot();
  });

  it("use-base-udir-3", () => {
    clearStore();
    registDir(dir1);
    registDir(dir2);
    registDir(dir3);

    const appConfig = resolveConfig("app");
    expect(appConfig.appid).toBe("project-3");
    expect(appConfig.version).toBe("20191118-1-2-3");
    expect(appConfig.nonHelper.payload).toBe("non-helper");
    const pathsConfig = resolveConfig("paths");
    expect(pathsConfig.entry).toBe("./src/index.js");
  });

  it("use-base-udir-2", () => {
    clearStore();
    registDir(dir1);
    registDir(dir2, false);
    registDir(dir3);

    const appConfig = resolveConfig("app");
    expect(appConfig.appid).toBe("project-2");
    expect(appConfig.version).toBe("20191118-1-3-2");
  });

  it("wait-deps", () => {
    const category = "wait-deps";
    setStoreExt("ts", category);
    registDir(dir1, category);
    registDir(dir2, category);
    registDir(dir3, false, category);

    const webpack = resolveConfig("webpack", category);
    const env = resolveConfig("env", category);
    expect(env.name).toBe("name-project-3");
    expect(env.remoteDev).toBe(true);
    expect(webpack.version).toBe("version-20191118-1-2-3");
    expect(webpack.remoteDev).toBe(true);
    expect(webpack.server.hot).toBe(true);
  });

  it("mutate", () => {
    const category = "mutate";
    setStoreExt("ts", category);
    registDir(dir1, category);
    registDir(dir2, category);
    registDir(dir3, false, category);

    const babel = resolveConfig("babel", category);

    expect(babel.presets.length).toBe(3);
    expect(babel.presets[1][0]).toBe("typescript");
    expect(babel.presetsLength).toBe(3);
    expect(babel.plugins.length).toBe(1);
    expect(babel.plugins[0][0]).toBe("transform-runtime");
  });

  it("use-value", () => {
    const category = "use-value";
    setStoreExt("ts", category);
    registDir(dir1, category);
    registDir(dir2, category);
    registDir(dir3, false, category);

    const useValue = resolveConfig("use-value", category);

    expect(useValue.baseDate).toBe("20191126");
    expect(useValue.version).toBe("20191118-1-2-3");
    expect(useValue.pluginName).toBe("transform-runtime");
  });
});
