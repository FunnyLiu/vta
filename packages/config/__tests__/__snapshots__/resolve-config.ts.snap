// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config hooks 1`] = `
"[
  \\"env-base-start\\",
  \\"app-base-start\\",
  \\"app-base-done\\",
  \\"app-user-start\\",
  \\"app-user-getted\\",
  \\"app-user-done\\",
  \\"env-base-done\\",
  \\"env-user-start\\",
  \\"env-user-getted\\",
  \\"env-user-done\\",
  \\"webpack-base-start\\",
  \\"webpack-base-done\\",
  \\"webpack-user-start\\",
  \\"webpack-server-base-start\\",
  \\"webpack-server-base-done\\",
  \\"webpack-server-user-start\\",
  \\"webpack-server-user-getted\\",
  \\"webpack-server-user-done\\",
  \\"webpack-user-getted\\",
  \\"webpack-user-done\\"
]"
`;
